### PR TITLE
[FLOC-3794] replace FlockerScriptTestsMixin should follow the make_foo_tests pattern

### PR DESCRIPTION
**Pull request by justnoise**
As requested in the Jira ticket, replaces inheritence based classes with a class generator.

### CODE REVIEW

File: flocker/testtools/__init__.py [hunk 1]
```diff
@@ -281,171 +281,180 @@ def __init__(self, argv=None):
...
 
-class StandardOptionsTestsMixin(object):
-    """Tests for classes decorated with ``flocker_standard_options``.
 
-    Tests for the standard options that should be available on every flocker
-    command.
+def make_standard_options_test(options):
+    """
+    Return a ``FlockerScriptTestCase`` which tests that the script
+    class provides ICommandLineScript
```
Comment (james-w) [inline flocker/testtools/__init__.py]: I'm not sure which "script class" this is referring to, it's not one of the
arguments to this function.

File: flocker/testtools/__init__.py [hunk 2]
```diff
@@ -281,171 +281,180 @@ def __init__(self, argv=None):
...
-    options = None
 
-    def test_sys_module_default(self):
+    :returns: a ``TestCase``
+    """
+    class StandardOptionsTestCase(TestCase):
         """
-        ``flocker_standard_options`` adds a ``_sys_module`` attribute which is
-        ``sys`` by default.
+        Test for classes that implement ICommandLineScript
```
Comment (james-w) [inline flocker/testtools/__init__.py]: I think this should perhaps be the docstring from the original Mixin?

i.e.

"Tests for the standard options that should be available on every flocker command."

File: flocker/cli/test/test_deploy_script.py [hunk 3]
```diff
@@ -6,33 +6,30 @@
...
-    script = DeployScript
-    options = DeployOptions
-    command_name = u'flocker-deploy'
 
 
 CONTROL_HOST = u"192.168.1.1"
 
 
-class DeployOptionsTests(StandardOptionsTestsMixin, SynchronousTestCase):
+class DeployOptionsTests(make_standard_options_test(DeployOptions)):
```
Comment (james-w) [inline flocker/cli/test/test_deploy_script.py]: :question: The factory returns a class for the standard interface tests. Perhaps the extra tests for the
behaviour of these specific options would be good as a separate class?

**Comment by james-w** [comment]
Hi,

Thanks for the branch, it looks great and is much cleaner.

There's a couple of docstrings that I think could be improved, and one
comment on the structure that you may want to use or to ignore.

I've asked jml to take a look at the branch as well as this is my first review.

Thanks,

James

File: flocker/testtools/__init__.py (hunk 2 above)
Comment (jml) [inline flocker/testtools/__init__.py]: If they also implement `ICommandLineScript`, then the docstring should also mention that (in addition to the original docstring).

Also, fwiw, use double backquotes to mark something as code in docstrings, e.g.

    ``ICommandLineScript``

**Comment by jml** [comment]
Thanks Brendan! It's great to see this code getting cleaned up.

There are merge conflicts (probably my fault, sorry). Please resolve them and make sure all tests pass before landing.

Thanks for the review @james-w, only a minor tweak suggested.

**Comment by justnoise** [comment]
Thanks for your comments James, I spaced a bit on the docstrings.  I'll get into resolving those merge conflicts.

**Comment by sarum90** [comment]
Updates look good. Just wait for greens on Jenkins and we can go ahead and merge.

**Pull request merged**